### ISSUE TITLE
Dispatcher Threads are leaked

### ISSUE THREAD

**Issue by sebthom** [eclipse/lsp4e#371]
When closing the last file of a project, the associated language server is shutdown.
However two dispatcher threads are not stopped. Repeatably closing/opening the same file then results in more and more dispatcher threads.

Here is an example after opening/closing the same JS file several times.

![image](https://user-images.githubusercontent.com/426959/212719113-b1074cfe-b910-422f-8825-489534562dfa.png)

The LanguageServerWrapper#stopDispatcher dispatcher method is only invoked via LanguageServiceAccessor#shutdownAllDispatchers once the LSP4E bundle gets stopped.

**Comment by ahmedneilhussain** [comment]
Hi - @mickaelistria  I was unaware that this code had been committed to master. I thought we were still waiting to decide between #344 and #347 and then polish whichever one up we decided was the better approach.

Did you cherry pick some bits for master? I wasn't aware...!

**Comment by ahmedneilhussain** [comment]
On closer inspection, it's not the dispatcher thread that is being leaked. It's the *launcher* threads. This is a longstanding issue (which I had spotted, and hadn't got round to mentioning to anyone). This are the threads created at https://github.com/eclipse/lsp4e/blob/4c79b78e52bad00acca50ac3e7d004bf0b43b59a/org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java#L291 which are used by the LSP4j to listen to LS responses. At the moment they get started from scratch every time the LSW starts the server and not cleaned up on LSW.shutdown, hence the leak. (Note that although we supply a new 'pool' executor, only one listener thread can ever get started from it by LSP4j).

It's a longstanding issue - if you run the test suite and monitor the threads from eclipse, you will see rather a lot of them by the time the suite finishes...!

**Comment by ahmedneilhussain** [comment]
The dispatcher is intentionally kept alive for the lifetime of the plugin, and reused if we restart the LS. That's by my design (debatably so, but by design nonetheless!). So although if you shutdown the LS and never reuse it from within your eclipse session, you will get an idle dispatcher thread hanging around, it won't be leaked.

**Comment by ahmedneilhussain** [comment]
Cleaning up the launcher threads is probably a good idea and should be fixed, but it wasn't introduced by https://github.com/eclipse/lsp4e/commit/5ea29bf033f6b07c2d9a9ea769ec0d83bde4f219 ...

Although I spot this all the time when running the tests from eclipse, it hasn't been a particular concern because we run our plugin in 'never time out and shut down' mode - obviously for plugins with a short idle timeout then it could be more of an issue.

**Comment by mickaelistria** [comment]
> Did you cherry pick some bits for master? I wasn't aware...!

Yes, it's what I tried to mention earlier when I said I was trying to merge things "commit after commit", in the sense that I would like to cherry-pick bits that are consensual enough to start getting them working and accumulate feedbachk and thus facilitate further discussions.

**Comment by ahmedneilhussain** [comment]
I've created #376 to try and improve this for testing (having investigated I think it's benign for production)
